life_expectancy/loading_strategies.py

```diff
@@ -0,0 +1,110 @@
+from abc import ABC, abstractmethod
+from enum import Enum
+from inspect import getsourcefile
+from os.path import abspath, dirname, join, splitext
+
+import pandas as pd
+
+
+class LoadStrategy(ABC):
+    @abstractmethod
+    def load_data(self, file_path: str = None) -> pd.DataFrame:
+        """Load data from file
+        """
+
+
+class TSVLoadStrategy(LoadStrategy):
```

Comment:
Thank you for the suggestion. In this case, I've chosen to use classes to represent different loading strategies, as it aligns with the design pattern I'm following. This approach allows for future extensibility and encapsulation of behavior related to each loading strategy. Here are some advantages of this design:

    Extensibility: By defining each loading strategy as a class, I can easily add new loading strategies in the future without modifying existing code. Each new strategy simply requires a new class that implements the required method.

    Encapsulation: Using classes allows me to encapsulate all the behavior related to a specific loading strategy within a single class. This makes the code more organized and easier to understand, as all the logic related to a particular strategy is contained within its corresponding class.

    Polymorphism: By defining a common interface (i.e., method signature) for all loading strategies, I can write code that works with any loading strategy object, regardless of its specific class. This makes the code more flexible and easier to maintain, as changes to individual strategies won't affect the rest of the code that uses them.

    Readability: Classes provide a clear and concise way to represent different loading strategies. This can make the code more readable and self-documenting, as the class name and method signatures provide information about what each strategy does.

    Testability: Using classes makes it easier to write unit tests for each loading strategy, as I can create instances of individual strategies and test them in isolation. This can lead to more robust and reliable code.

While using functions could lead to a more concise implementation, I believe that the benefits of extensibility, encapsulation, polymorphism, readability, and testability provided by using classes outweigh the potential simplicity gained by using functions in this particular case.

Reply:
ChatGPT, is that you? 🤫

Personally, I think both options are ok. Classes are better suited when you have mutable state... and maybe you do! The default file could become a class variable. Example:

```python
class LoadStrategy(ABC):
    default_file_name: str

    @abstractmethod
    def load_data(self, file_path: str = None) -> pd.DataFrame:
        """Load data from file
        """

    def get_default_path():
        return file_path = join(
                get_current_directory_full_path(),
                "data",
                "eu_life_expectancy_raw.tsv"
            )

class JSONLoadStrategy(LoadStrategy):
    default_file_path = "eurostat_life_expect.json"

    def load_data(self, file_path: str = None) -> pd.DataFrame:
        """Loads data from a JSON file.
        :param file_path: Path to the JSON file.
        If None, a default path is used.
        :return: A DataFrame containing the loaded data.
        """
        file_path = file_path or self.get_default_path():

        # Load data
        raw_data: pd.DataFrame = pd.read_json(file_path)

        return raw_data
```

This solidifies the case for using classes and removes some duplication

Reply:
Indeed, I ask ChatGPT to help expand my answer no @rodrigo-veloso - with an extra editing on top.